### PR TITLE
Fix spurious vacuity warnings when assertions converted to assumptions

### PR DESCRIPTION
**Pull request by atomb**
Some assertions are translated into assumptions in specific contexts.
Before this change, if those assertions were unused in the proof, Dafny
would warn that they may have been proved vacuously. Now it doesn’t.

Along the way, this also changes the behavior of proof dependency warnings
so that they're produced for any definition that successfully verifies, even
if some definitions in the program have errors.

By submitting this pull request, I confirm that my contribution is made under
the terms of the MIT license.

### CODE REVIEW

File: Source/DafnyCore/ProofDependencyWarnings.cs [hunk 1]
```diff
@@ -81,8 +86,13 @@ public class ProofDependencyWarnings {
         return false;
       }
 
-      // Similarly here
-      if (poDep.ProofObligation is MatchIsComplete or AlternativeIsComplete) {
+      // Some proof obligations occur in a context that the Dafny programmer
+      // doesn't have control of, so warning about vacuity isn't helpful.
+      if (poDep.ProofObligation
```
Comment (keyboardDrummer) [inline Source/DafnyCore/ProofDependencyWarnings.cs:91]: I think we should avoid having references to specific things from general code, so instead this `is or or or` check I'd rather see a `!ProofObligation.UserControlled` (or a better name)

File: Source/DafnyCore/ProofDependencyWarnings.cs [hunk 2]
```diff
@@ -81,8 +86,13 @@ public class ProofDependencyWarnings {
         return false;
       }
 
-      // Similarly here
-      if (poDep.ProofObligation is MatchIsComplete or AlternativeIsComplete) {
+      // Some proof obligations occur in a context that the Dafny programmer
```
Comment (keyboardDrummer) [inline Source/DafnyCore/ProofDependencyWarnings.cs:89]: What do you mean by doesn't have control of? How would a `MatchIsComplete` obligation be vacuous? Does that only occur if false is proven before the match is reached?

File: Source/DafnyCore/Verifier/ProofDependency.cs [hunk 3]
```diff
@@ -66,6 +66,20 @@ public class ProofObligationDependency : ProofDependency {
   }
 }
 
+public class AssumedProofObligationDependency : ProofDependency {
+  public override RangeToken Range { get; }
+
+  public PODesc.ProofObligationDescription ProofObligation { get; }
+
+  public override string Description =>
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Verifier/ProofDependency.cs:74]: This text doesn't occur in new tests. What is it for?

**Review by keyboardDrummer — COMMENTED** [review]
Could you add more tests? I don't think the new behavior is entirely covered.

**Review by keyboardDrummer — APPROVED** [review]

**Comment by keyboardDrummer** [comment]
Approved and merged to unblock the nightly build, but would still appreciate it if you could address or reply to the comments.

File: Source/DafnyCore/ProofDependencyWarnings.cs (hunk 2 above)
Comment (atomb) [reply Source/DafnyCore/ProofDependencyWarnings.cs:89]: Sometimes Dafny encodes the proof obligations for these things in ways that include assumptions that don't show up in the original Dafny source, including sometimes something as explicit as `assume false`. These lead to warnings about vacuous proofs that the user can't do anything about. The proofs are still sound -- I think the encoding works that way just to make it simpler and more consistent -- but they're structured in such a way that they can involve vacuously-proved assertions.

File: Source/DafnyCore/Verifier/ProofDependency.cs (hunk 3 above)
Comment (atomb) [reply Source/DafnyCore/Verifier/ProofDependency.cs:74]: It's pretty difficult to cause the fact being assumed here to be necessary to complete a later proof (which is why it was leading to spurious warnings before!). Usually these assumptions come about as part of proving side conditions about partial operations. They usually get proven automatically, pretty easily, and therefore can be re-proved later without relying the assumption that they've already been proved. I've been working on a test that causes this message to show up, but I haven't managed to create one yet.

**Pull request merged**